processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,385 @@
+      final AggregatorFactory[] aggregatorFactories,
+      final long timeoutMs
+  )
+  {
+    this.bufferSupplier = bufferSupplier;
+    this.keySerde = keySerde;
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+
+    this.keySize = keySerde.keySize();
+    int offset = keySize;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+    this.recordSize = offset;
+    this.timeoutMs = timeoutMs;
+  }
+
+  @Override
+  public void init()
+  {
+    if (!initialized) {
+      buffer = bufferSupplier.get();
+      maxSlotNum = buffer.capacity() / recordSize;
+      Preconditions.checkState(
+          maxSlotNum > 2,
+          "Buffer[%s] should be large enough to store at least three records[%s]",
+          buffer.capacity(),
+          recordSize
+      );
+
+      reset();
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key, int notUsed)
+  {
+    return aggregate(key);
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    try {
+      final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+
+      if (keyBuffer.remaining() != keySize) {
+        throw new IAE(
+            "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+            keyBuffer.remaining(),
+            keySize
+        );
+      }
+
+      final int prevRecordOffset = curWriteIndex * recordSize;
+      if (curWriteIndex == -1 || !keyEquals(keyBuffer, buffer, prevRecordOffset)) {
+        initNewSlot(keyBuffer);
+      }
+
+      final int curRecordOffset = curWriteIndex * recordSize;
+      for (int i = 0; i < aggregatorOffsets.length; i++) {
+        aggregators[i].aggregate(buffer, curRecordOffset + aggregatorOffsets[i]);
+      }
+
+      return AggregateResult.ok();
+    }
+    catch (Throwable t) {
+      finish();
+      throw t;
+    }
+  }
+
+  private boolean keyEquals(ByteBuffer curKeyBuffer, ByteBuffer buffer, int bufferOffset)
+  {
+    int i = 0;
+    for (; i + Long.BYTES <= keySize; i += Long.BYTES) {
+      if (curKeyBuffer.getLong(i) != buffer.getLong(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    for (; i + Integer.BYTES <= keySize; i += Integer.BYTES) {
+      if (curKeyBuffer.getInt(i) != buffer.getInt(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    for (; i < keySize; i++) {
+      if (curKeyBuffer.get(i) != buffer.get(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private void initNewSlot(ByteBuffer newKey)
+  {
+    increaseWriteIndex();
```

Comment:
Changing volatile variable before updating the buffer data doesn't ensure correct visibility

Reply:
Would you elaborate more on ensuring correct visibility? It's designed that a reading thread cannot read while a writing thread is writing to a slot in the buffer. It means, once ```curWriteIndex``` is increased, the ```curWriteIndex - 1```th slot is safe to be read until it is overwritten.